### PR TITLE
Sarah afiya/feature/llm implementation

### PR DESCRIPTION
**Pull request by sarah-afiya**
The last 5 commits are of this branch (commits from Feb 15)

### CODE REVIEW

File: documentation_quality_analysis/analyze_library/prioritize_api/llm/detect_api_discussion.py [hunk 1]
```diff
@@ -11,19 +12,25 @@
...
-    response = client.chat.completions.create(
-        model=model,
-        messages=messages,
-        temperature=0
-    )
+    completion = client.chat.completions.create(model=model,
+    messages=[
+        {
+            "role": "user",
+            "content": "\n\nThis is a test!",
```
Comment (snadi) [inline documentation_quality_analysis/analyze_library/prioritize_api/llm/detect_api_discussion.py]: are you intentionally leaving "this is a test" in the code?

File: documentation_quality_analysis/analyze_library/prioritize_api/llm/data/SO_query_examples_set2.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    "api": "pandas.pivot_table",
+    "discusses_api": "no"
+  },
+  "example2": {
+    "post": "python pandas: pivot_table silently drops indices with nans:\n\nIs there an option not to drop the indices with NaN in them? I think silently dropping these rows from the pivot will at some point cause someone serious pain.\n\nimport pandas\nimport numpy\n\na = [['a', 'b', 12, 12, 12], ['a', numpy.nan, 12.3, 233., 12], ['b', 'a', 123.23, 123, 1], ['a', 'b', 1, 1, 1.]]\n\ndf = pandas.DataFrame(a, columns=['a', 'b', 'c', 'd', 'e'])\n\ndf_pivot = df.pivot_table(index=['a', 'b'], values=['c', 'd', 'e'], aggfunc=sum)\nprint(df)\nprint(df_pivot)\nOutput:\n\n   a    b       c    d   e\n0  a    b   12.00   12  12\n1  a  NaN   12.30  233  12\n2  b    a  123.23  123   1\n3  a    b    1.00    1   1\n          c    d   e\na b\na b   13.00   13  13\nb a  123.23  123   1",
+    "api": "pandas.pivot_table",
+    "discusses_api": "yes"
+  }
+}
```
Comment (snadi) [inline documentation_quality_analysis/analyze_library/prioritize_api/llm/data/SO_query_examples_set2.json:12]: so what is the difference between set1 and set2? and why are there separate .txt files with examples? Are these meant to be used in the json files?

File: documentation_quality_analysis/analyze_library/prioritize_api/llm/messages/few_shot_learning.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    {"role": "assistant",
+    "content":  "$answer1"},
+    {"role": "user",
+    "content":  "This is a stack overflow post:\n\nPost: \"\"\"\n'$example2'\n\"\"\"\n\nIs the main topic of this post this API: '$api2'?\n\nAnswer in yes or no"},
+    {"role": "assistant",
+    "content":  "$answer2"},
+    {"role": "user",
+    "content":  "This is a stack overflow post:\n\nPost: \"\"\"\n'$post_body'\n\"\"\"\n\nIs the main topic of this post this API: '$api_fqn'?\n\nAnswer in yes or no"}
+
+  ]
```
Comment (snadi) [inline documentation_quality_analysis/analyze_library/prioritize_api/llm/messages/few_shot_learning.json:14]: template makes sense.

File: documentation_quality_analysis/analyze_library/prioritize_api/llm/messages/few_shot_learning_2.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "messages": [
+    {
+      "role": "user",
+      "content": ""
+    }
+  ]
```
Comment (snadi) [inline documentation_quality_analysis/analyze_library/prioritize_api/llm/messages/few_shot_learning_2.json:7]: what is this second template?

File: documentation_quality_analysis/analyze_library/prioritize_api/llm/templates/template_2.txt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+"""
+Is the main topic of this post this API: '$api2'?
+"""
+
+The correct answer is: '$answer2'
+
+Now, this is another Stack Overflow post:
+‘$post_body’
+
+Is the main topic of this post this API: '$api_fqn'?
```
Comment (snadi) [inline documentation_quality_analysis/analyze_library/prioritize_api/llm/templates/template_2.txt:30]: so you are doing zero shot and few shot, in both prompting styles? (json messages conversation vs. a single prompt conversation?)

File: documentation_quality_analysis/analyze_library/prioritize_api/llm/discusses_api_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+def get_prompt_messages(library, api_fqn, post_body, mode: Modes):
+    metadata: dict = {
+        'library': library,
+        'api_fqn': api_fqn,
+        'post_body': post_body
+    }
+
+    messages = []
+
+    if mode == Modes.ZERO_SHOT:
```
Comment (snadi) [inline documentation_quality_analysis/analyze_library/prioritize_api/llm/discusses_api_pipeline.py:26]: Modes --> Mode

File: documentation_quality_analysis/analyze_library/prioritize_api/llm/dtos/modes.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+from enum import Enum
+
+
+class Modes(Enum):
```
Comment (snadi) [inline documentation_quality_analysis/analyze_library/prioritize_api/llm/dtos/modes.py:4]: `class PromptMode(Enum)`

**Review by snadi — COMMENTED** [review]
Looks good (as far as I can tell since reviewing a subset of commits is confusing)

**Comment by sarah-afiya** [comment]
@snadi I addressed your feedback from here and also refactored the templates (now reading the templates from the template directory rather than having them hard-coded).  

I somehow messed up commits while trying to rebase, I think I fixed it now, you should see only the file changes and commits related to LLM implementation here now. Could you please give it another look and approve if it's okay?

**Review by snadi — APPROVED** [review]
Quickly skimmed through it and I think it matches what we discussed.

**Comment by snadi** [comment]
and yes, I can see only the commits related to the LLM evaluation

**Pull request merged**